### DISCUSSION TITLE
Rendering an outline box for specific blocks

### DISCUSSION THREAD

**Discussion by anarchy1337x** [FabricMC/fabric-api#2722] [Mod Dev Support]
# Error

```
FATAL ERROR in native method: Thread[Render thread,10,main]: No context is current or a function that is not available in the current context was called. The JVM will abort execution.
	at org.lwjgl.opengl.GL11.glPushMatrix(Native Method)
	at me.anarchy.demo.demo.client.demoClient.lambda$onInitializeClient$0(demoClient.java:75)
	at me.anarchy.demo.demo.client.demoClient$$Lambda$3772/0x0000000801535068.beforeBlockOutline(Unknown Source)
	at net.minecraft.client.render.WorldRenderer.handler$zhm000$beforeRenderOutline(WorldRenderer.java:3496)
	at net.minecraft.client.render.WorldRenderer.render(WorldRenderer.java:1168)
	at net.minecraft.client.render.GameRenderer.renderWorld(GameRenderer.java:1024)
	at net.minecraft.client.render.GameRenderer.render(GameRenderer.java:833)
	at net.minecraft.client.MinecraftClient.render(MinecraftClient.java:1102)
	at net.minecraft.client.MinecraftClient.run(MinecraftClient.java:752)
	at net.minecraft.client.main.Main.main(Main.java:220)
	at net.minecraft.client.main.Main.main(Main.java:56)
	at java.lang.invoke.LambdaForm$DMH/0x0000000800e18000.invokeStaticInit(java.base@17.0.3/LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/0x0000000800c01800.invokeExact_MT(java.base@17.0.3/LambdaForm$MH)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:461)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
	at java.lang.invoke.LambdaForm$DMH/0x0000000800c01000.invokeStatic(java.base@17.0.3/LambdaForm$DMH)
	at java.lang.invoke.LambdaForm$MH/0x0000000800c01800.invokeExact_MT(java.base@17.0.3/LambdaForm$MH)
	at net.fabricmc.devlaunchinjector.Main.main(Main.java:86)l
```

# Context

I'm trying to render outlines of specific blocks that are visible through walls, i've tried multiple times to use the raw opengl libraries, but i get the same issue everytime, no context specified. I've tried looking through the documentation, and various forge-alternatives. I've attempted to use pre-made libraries to render the boxes, it does it successfully but the offset is not at the block, and doesn't change no matter where i move.

I've been researching and apparently i need to use the tesselator and the built-in bufferbuilder, i'd love some guidance (and|or) clarification on what i should use.

Thanks!

```java
GL11.glEnable(GL11.GL_BLEND);
GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
GL11.glEnable(GL11.GL_LINE_SMOOTH);
GL11.glLineWidth(2);
GL11.glDisable(GL11.GL_TEXTURE_2D);
GL11.glEnable(GL11.GL_CULL_FACE);
GL11.glDisable(GL11.GL_DEPTH_TEST);

double renderPosX = instance.gameRenderer.getCamera().getPos().getX();
double renderPosY = instance.gameRenderer.getCamera().getPos().getY();
double renderPosZ = instance.gameRenderer.getCamera().getPos().getZ();

GL11.glPushMatrix();
GL11.glTranslated(-renderPosX, -renderPosY, -renderPosZ);

GL11.glColor4f(0, 1, 0, 0.25F);

for(BlockEntity blockEntity : getBlockEntities()) {
	assert instance.world != null;

	Block block = instance.world.getBlockState(blockEntity.getPos()).getBlock();
	if(block.getTranslationKey().endsWith("chest")) {
		BlockPos blockPosition = blockEntity.getPos();

		int x = blockPosition.getX();
		int y = blockPosition.getX();
		int z = blockPosition.getX();
		Box box = new Box(x, y, z, x + 1, y + 1, z + 1);

		RenderUtils.drawOutlinedBox(box, matrixStack);
	}
}

GL11.glPopMatrix();

GL11.glEnable(GL11.GL_DEPTH_TEST);
GL11.glEnable(GL11.GL_TEXTURE_2D);
GL11.glDisable(GL11.GL_BLEND);
GL11.glDisable(GL11.GL_LINE_SMOOTH);
```

# drawOutlinedBox
```java
public static void drawOutlinedBox(Box bb, MatrixStack matrixStack) {
	Matrix4f matrix = matrixStack.peek().getPositionMatrix();
	Tessellator tessellator = RenderSystem.renderThreadTesselator();
	BufferBuilder bufferBuilder = tessellator.getBuffer();
	RenderSystem.setShader(GameRenderer::getPositionShader);

	bufferBuilder.begin(VertexFormat.DrawMode.DEBUG_LINES, VertexFormats.POSITION);
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.minY, (float)bb.minZ).next();
	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.minY, (float)bb.minZ).next();

	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.minY, (float)bb.minZ).next();
	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.minY, (float)bb.maxZ).next();

	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.minY, (float)bb.maxZ).next();
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.minY, (float)bb.maxZ).next();

	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.minY, (float)bb.maxZ).next();
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.minY, (float)bb.minZ).next();

	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.minY, (float)bb.minZ).next();
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.maxY, (float)bb.minZ).next();

	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.minY, (float)bb.minZ).next();
	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.maxY, (float)bb.minZ).next();

	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.minY, (float)bb.maxZ).next();
	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.maxY, (float)bb.maxZ).next();

	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.minY, (float)bb.maxZ).next();
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.maxY, (float)bb.maxZ).next();

	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.maxY, (float)bb.minZ).next();
	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.maxY, (float)bb.minZ).next();

	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.maxY, (float)bb.minZ).next();
	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.maxY, (float)bb.maxZ).next();

	bufferBuilder.vertex(matrix, (float)bb.maxX, (float)bb.maxY, (float)bb.maxZ).next();
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.maxY, (float)bb.maxZ).next();

	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.maxY, (float)bb.maxZ).next();
	bufferBuilder.vertex(matrix, (float)bb.minX, (float)bb.maxY, (float)bb.minZ).next();
	tessellator.draw();
}
```

**Comment by warjort** [comment]
I'm pretty sure you shouldn't be using the OpenGl library directly.

This is not something I know a lot about, but maybe look at
````WorldRenderer.draw(Cuboid)ShapeOutline```` or ````WorldRenderer.drawBox()````
and where they get used.
The VoxelShape.forEachEdge() is probably what you want to know what vertices to draw. 
With the shape coming from the current block state.

IIRC the co-ordinate system is at the camera when drawing block outlines so you need to translate by
blockPos - cameraPos

**Reply by anarchy1337x** [reply]
Hey there thanks for the reply,

i'm still encountering some issues when trying to translate the coordinates. I'm successfully rendering the block, but it appears that it's not properly getting the camera as it's offset by a huge margin, and it seems to be only relative to the center of the world.

I've attempted to use the WorldRenderer, however it does not the render the box.

![image](https://user-images.githubusercontent.com/49851060/205467899-a6ddb58e-dc4e-4aab-a6eb-8447771fc7fc.png)

As for the code i'm using, i've attached it below.

I'd love to know what i'm doing wrong, thanks!


```java
BlockPos blockPosition = blockEntity.getPos();
int x = blockEntity.getPos().getX();
int y = blockEntity.getPos().getX();
int z = blockEntity.getPos().getX();

double camX = c.camera().getPos().getX();
double camY = c.camera().getPos().getY();
double camZ = c.camera().getPos().getZ();

VertexConsumer consumer = c.consumers().getBuffer(RenderLayer.getLines());
RenderUtils.drawOutlinedBox(new Box(x - camX, y - camY, z - camZ, x + 1 - camX, y + 1 - camY, z + 1 - camZ), matrixStack);
```

**Comment by warjort** [comment]
````
int x = blockEntity.getPos().getX();
int y = blockEntity.getPos().getX(); <-- y
int z = blockEntity.getPos().getX(); <-- z
````

Beyond that, it's pretty much impossible to help you debug code if you only snippets out of context.
e.g. where/when is this code called so we can understand what the value of the transformation matrix is on top of the matrix stack

**Reply by anarchy1337x** [reply]
I have no clue myself, and about the xyz, it was initally correct before, i had to swap out the camera variables and forgot to change them back, it didn't solve my issue.

As for the transformation layers the function i attached above is all the things it does to the tesselator before and after, however i still have no clue why the built in (fabric/minecraft) function doesn't work.

If it gives any clarification i'm running the code in the ```BEFORE_BLOCK_OUTLINE``` event. Also i'm pretty sure i'm not getting the vertex consumer properly, could you confirm this?

Thanks!

**Reply by anarchy1337x** [reply]
extra - the code is being run in the ``BEFORE_BLOCK_OUTLINE`` event, and it's in the main initializer for the mod.

**Reply by anarchy1337x** [reply]
I'm not too sure which event, or which event i should hook with a mixin as there's barely any documentation on the fabric wiki to do so.

**Reply by anarchy1337x** [reply]
I've got the rendering working now, but it appears that the alpha (opacity) isn't rendering correctly, any reason why?

![image](https://user-images.githubusercontent.com/49851060/205472113-eea5d637-9d7c-4ab6-9bbc-17c7e752dca3.png)

**Reply by warjort** [reply]
Like I said before, you need to post the full code, not snippets.

Guess: You are probably passing the rgba to the vertext consumer in whatever RenderUtils does?

**Comment by moist-socks** [comment]
The think the problem is that you are not translating your box correctly. To fix this you could add a new function to your RenderUtils class.
`
public static void drawTranslatedBox(Vec3d boxPos, WorldRenderContext context) {
        // Size of box
        Box box = new Box(-0.5, -0.5, -0.5, 0.5, 0.5, 0.5);
        MatrixStack matrixStack = context.matrixStack();

        // Center position
        Vec3d centerPos = boxPos.add(0.5, 0.5, 0.5);

        // Get the postive relative to the camera
        Vec3d cameraPos = context.camera().getPos();
        Vec3d relBoxPos = centerPos.subtract(cameraPos);

        // Translate matrix stack to fixed position
        matrixStack.push();
        matrixStack.translate(relBoxPos.x, relBoxPos.y, relBoxPos.z);

        // Draw the box
        RenderUtils.drawOutlinedBox(box, matrixStack);
        
        matrixStack.pop();
    }
`
This function creates a new stack of matrices. If this is not here it will translate other matrices than just your box. The function translates the matrices relative to your camera to give the box a fixed position in your world. Last, it pops the matrix to remove the new matrix stack layer.

To run you this code you could do something like this.
`
// Register event listener
WorldRenderEvents.BEFORE_ENTITIES.register((WorldRenderContext context) -> {
   // Draw box
   RenderUtils.drawTranslatedBox(boxPos, context);
});
`
Beware, this callback runs every frame during the entity rendering pass.
To make work with what you are using you can make a for loop that passes through all block entities that are chest. Then, add them to an array. After that you can iterate through the array drawing the box for every coordinate.
